cvmfs/util/capabilities.cc

```diff
@@ -0,0 +1,268 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+
+#include <errno.h>
+
+#include <cassert>
+
+#ifdef __APPLE__
+#include <unistd.h>
+#endif
+
+#include "util/capabilities.h"
+#include "util/logging.h"
+#include "util/platform.h"
+#include "util/posix.h"
+
+#ifdef CVMFS_NAMESPACE_GUARD
+namespace CVMFS_NAMESPACE_GUARD {
+#endif
+
+#ifdef __APPLE__
+
+bool ClearPermittedCapabilities(int,
+                                const cap_value_t *,
+                                int,
+                                const cap_value_t *) {
+  return true;
+}
+
+namespace {
+
+bool obtainCapability(const cap_value_t, const char *) {
+  // there are no individual capabilities on OSX so switch back to root
+  return (SwitchCredentials(0, getgid(), true));
+}
+
+bool checkCapabilityPermitted(const cap_value_t) {
+  return (getuid() == 0);
+}
+
+} // namespace
+
+#else
+
+#include <sys/prctl.h>
+
+// Clear all CAP_PERMITTED capabilities except those requested.
+// This function requires being run with CAP_SETPCAP capability permitted.
+// If the real uid & gid do not match the effective uid & gid, it also
+// requires CAP_SETUID and CAP_SETGID capabilities to be permitted and
+// ends up switching the real uid & gid to match the incoming effective
+// uid & gid.  Beware that switching the uid is not thread-safe; it is
+// process-wide and clears all capabilities from threads that do not
+// have keepcaps enabled.
+// Capabilities listed in inheritcaps will be made inheritable.
+
+bool ClearPermittedCapabilities(int nreservecaps,
+                                const cap_value_t *reservecaps,
+                                int ninheritcaps,
+                                const cap_value_t *inheritcaps) {
+  int retval = 0;
+  uid_t uid, gid;
+
+  if (!SetpcapCapabilityPermitted()) {
+    if (nreservecaps > 0) {
+      LogCvmfs(kLogCvmfs, kLogDebug,
+        "Capabilities cannot be reserved because setpcap is not permitted.");
+      return false;
+    }
+    LogCvmfs(kLogCvmfs, kLogDebug,
+      "Capabilities are already cleared because setpcap is not permitted.");
+    return true;
+  }
+  
+  uid = geteuid();
+  gid = getegid();
+  if ((uid != getuid()) || (gid != getgid())) {
+    // Only do setuid & setgid when necessary because it is a process-wide
+    // setting, not a per-thread setting.
+    if (!ObtainSetuidgidCapabilities()) {
+      LogCvmfs(kLogCvmfs, kLogStderr | kLogDebug,
```

Comment:
Since this function can be executed after `Deamonize()`, we should not log to stderr but only to syslog.
```suggestion
      LogCvmfs(kLogCvmfs, kLogSyslogErr | kLogDebug,
```